### PR TITLE
updated rebate to use native denom instead of reward denom

### PR DESCRIPTION
**Pull request by sampocs**
## Context
The rebate must now be issued using the native denom (DYDX) instead of the reward denom (USDC). 

## Brief Changelog
* Removed `CalculateRewardsSplitBeforeRebate` and `CalculateRewardsSplitAfterRebate` in favor of a simpler function `CalculateRewardsSplit` that returns the portion of rebate, stride fee, and reinvestment
* Updated `withdrawal-rewards` callback (the callback on the USDC query) to trade the full queried amount instead of checking for a rebate
* Updated `withdrawal-host` callback (the callback on the DYDX query) to optionally fund the community pool if a rebate was specified
* Updated unit tests

### CODE REVIEW

File: x/stakeibc/keeper/reward_converter.go [hunk 1]
```diff
@@ -53,130 +59,73 @@ type ForwardMetadata struct {
...
-
-		// It shouldn't be possible to have 0 token supply (since there are rewards and there was a community pool stake)
-		// This will also prevent a division by 0 error
-		if stTokenSupply.IsZero() {
-			return sdkmath.ZeroInt(), sdkmath.ZeroInt(), errorsmod.Wrapf(types.ErrDivisionByZero,
-				"unable to calculate rebate amount for %s since total delegations are 0", hostZone.ChainId)
-		}
-
-		// It also shouldn't be possible for the liquid stake amount to be greater than the full TVL
-		if rebateInfo.LiquidStakedStTokenAmount.GT(stTokenSupply) {
```
Comment (riley-stride) [inline x/stakeibc/keeper/reward_converter.go:158]: Should we keep these sanity checks? And some of the others e.g. `stTokenSupply.IsZero()`

File: x/stakeibc/keeper/reward_converter.go [hunk 2]
```diff
@@ -53,130 +59,73 @@ type ForwardMetadata struct {
...
-	hostZone types.HostZone,
-	rewardsAmount sdkmath.Int,
-) (strideFeeAmount sdkmath.Int, reinvestAmount sdkmath.Int, err error) {
-	// Get the fee rate and total fees from params (e.g. 0.1 for 10% fee)
-	strideFee := sdk.NewIntFromUint64(k.GetParams(ctx).StrideCommission)
-	totalFeeRate := sdk.NewDecFromInt(strideFee).Quo(sdk.NewDec(100))
-
-	// Check if the chain has a rebate
-	rebateInfo, chainHasRebate := hostZone.SafelyGetCommunityPoolRebate()
+	rebateAmount := sdk.NewDecFromInt(totalFeesAmount).Mul(contributionRate).Mul(rebateInfo.RebateRate).TruncateInt()
```
Comment (riley-stride) [inline x/stakeibc/keeper/reward_converter.go:119]: Logic seems sound

File: x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go [hunk 3]
```diff
@@ -59,33 +55,14 @@ func WithdrawalRewardBalanceCallback(k Keeper, ctx sdk.Context, args []byte, que
...
-
-		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_WithdrawalRewardBalance,
-			"Sending rebate tokens %v %s to community pool",
-			rebateAmount, tradeRoute.RewardDenomOnRewardZone))
-	}
-
-	// Transfer the amount leftover after to the rebate to the trade zone so it can be swapped for the native token
-	// We transfer both the amount to be reinvested, and the amount for the stride fee
-	if err := k.TransferRewardTokensHostToTrade(ctx, tradeAmount, tradeRoute); err != nil {
+	// Transfer the reward amount to the trade zone so it can be swapped for the native token
```
Comment (riley-stride) [inline x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go:58]: Should we check for 0 amount? I see `UnmarshalAmountFromBalanceQuery` handles the case where the acct is uninitialized, but not clear it handles the 0 balance case - we prob don't want to send a `TransferRewardTokensHostToTrade` if the bal is 0.

**Review by riley-stride — COMMENTED** [review]
Looking solid to me - great work!

File: x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go [hunk 4]
```diff
@@ -25,10 +25,6 @@ func WithdrawalRewardBalanceCallback(k Keeper, ctx sdk.Context, args []byte, que
 		"Starting withdrawal reward balance callback, QueryId: %vs, QueryType: %s, Connection: %s", query.Id, query.QueryType, query.ConnectionId))
 
 	chainId := query.ChainId
-	hostZone, err := k.GetActiveHostZone(ctx, chainId)
```
Comment (asalzmann) [inline x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go:28]: Why remove this?

**Review by asalzmann — APPROVED** [review]
lgtm - although I'm slightly less familiar with this code, should get a review from @shellvish and @ethan-stride too!

**Review by ethan-stride — APPROVED** [review]
LGTM

One weird thing is that StrideCommission and RebateRate are not stored in the same format.  We might also want to rename StrideCommission if it is really the percent of total fees not just the stride fees.

File: x/stakeibc/keeper/reward_converter.go (hunk 1 above)
Comment (sampocs) [reply x/stakeibc/keeper/reward_converter.go:158]: they are [still in there](https://github.com/Stride-Labs/stride/pull/1162/files/3ac87f999fb270a059ad67c22df73a23b4f529c1#diff-8850ef830f8cbe0e6ca0be787be1036bc793215c526fb43491a812ca4f0bb079R105-R114). 

They were previously in both the BeforeRebate and AfterRebate functions. But now there's only one function.

File: x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go (hunk 3 above)
Comment (sampocs) [reply x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go:58]: `TransferRewardTokensHostToTrade` doesn't send a transfer if the amount is less than the min configured amount

File: x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go (hunk 4 above)
Comment (sampocs) [reply x/stakeibc/keeper/icqcallbacks_withdrawal_reward_balance.go:28]: the hostZone struct is not used anymore in this function

**Review by riley-stride — APPROVED** [review]

**Review by shellvish — APPROVED** [review]
This looks great, thanks a lot for doing this! Seems like this _greatly_ simplifies the rebate logic!

File: x/stakeibc/keeper/icqcallbacks_withdrawal_host_balance.go [hunk 5]
```diff
@@ -59,17 +59,18 @@ func WithdrawalHostBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query
...
 	// Prepare MsgSends from the withdrawal account
-	feeCoin := sdk.NewCoin(hostZone.HostDenom, feeAmount)
-	reinvestCoin := sdk.NewCoin(hostZone.HostDenom, reinvestAmount)
+	feeCoin := sdk.NewCoin(hostZone.HostDenom, rewardsSplit.StrideFeeAmount)
+	reinvestCoin := sdk.NewCoin(hostZone.HostDenom, rewardsSplit.ReinvestAmount)
+	rebateCoin := sdk.NewCoin(hostZone.HostDenom, rewardsSplit.RebateAmount)
 
 	var msgs []proto.Message
-	if feeCoin.Amount.GT(sdk.ZeroInt()) {
+	if feeCoin.Amount.GT(sdkmath.ZeroInt()) {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icqcallbacks_withdrawal_host_balance.go:73]: `sdk` is the deprecated one!

File: x/stakeibc/keeper/reward_converter.go [hunk 6]
```diff
@@ -53,130 +59,73 @@ type ForwardMetadata struct {
...
-
 	// Get supply of stTokens to determine the portion of TVL that the community pool liquid stake makes up
 	stDenom := utils.StAssetDenomFromHostZoneDenom(hostZone.HostDenom)
 	stTokenSupply := k.bankKeeper.GetSupply(ctx, stDenom).Amount
 
 	// It shouldn't be possible to have 0 token supply (since there are rewards and there was a community pool stake)
 	// This will also prevent a division by 0 error
 	if stTokenSupply.IsZero() {
-		return sdkmath.ZeroInt(), sdkmath.ZeroInt(), errorsmod.Wrapf(types.ErrDivisionByZero,
+		return rewardSplit, errorsmod.Wrapf(types.ErrDivisionByZero,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/reward_converter.go:106]: this is a new function with a new return type - the diffs just look weird cause it overlapped the old functions a lot

File: x/stakeibc/keeper/icqcallbacks_withdrawal_host_balance.go [hunk 7]
```diff
@@ -87,6 +88,15 @@ func WithdrawalHostBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query
 		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_WithdrawalHostBalance,
 			"Preparing MsgSends of %v from the withdrawal account to the delegation account (for reinvestment)", reinvestCoin.String()))
 	}
+	if rebateCoin.Amount.GT(sdkmath.ZeroInt()) {
+		fundMsg, err := k.BuildFundCommunityPoolMsg(ctx, hostZone, sdk.NewCoins(rebateCoin), types.ICAAccountType_WITHDRAWAL)
+		if err != nil {
+			return err
+		}
+		msgs = append(msgs, fundMsg...)
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_WithdrawalHostBalance,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icqcallbacks_withdrawal_host_balance.go:97]: I considered the same thing, but we don't have access to that info outside of the function. I could move the log into the `BuildFundCommunityPoolMsg` function, it wont have the ICQCallback log prefix though (so will be unaesthetic) but probably still better?

[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icqcallbacks_withdrawal_host_balance.go]: moved log [here](https://github.com/Stride-Labs/stride/pull/1162/commits/0f83ab1338acb7ce617485a211a3ea346bd041cc) - resolving!

**Pull request merged**